### PR TITLE
ARROW-5876: [C++][Python] add basic auth flight proto message to C++ and Python 

### PR DESCRIPTION
**Pull request by rymurr**

### CODE REVIEW

File: python/pyarrow/_flight.pyx [hunk 1]
```diff
@@ -211,6 +211,47 @@ cdef class Result:
...
+    def username(self):
+        """Get the username."""
+        return self.basic_auth.get().username
+
+    @property
+    def password(self):
+        """Get the password."""
+        return self.basic_auth.get().password
+
+    @staticmethod
```
Comment (lihalite) [inline python/pyarrow/_flight.pyx]: Maybe for consistency with the equivalent FlightInfo, etc. methods we can call these `serialize` and `deserialize`?

Comment (rymurr) [reply python/pyarrow/_flight.pyx]: agree, fixed

File: cpp/src/arrow/python/flight.cc [hunk 2]
```diff
@@ -245,6 +245,19 @@ Status CreateFlightInfo(const std::shared_ptr<arrow::Schema>& schema,
...
 }
 
+Status DeserializeBasicAuth(const std::string& buf,
+                            std::unique_ptr<arrow::flight::BasicAuth>* out) {
+    auto basic_auth = new arrow::flight::BasicAuth();
+    *out = std::unique_ptr<arrow::flight::BasicAuth>(basic_auth);
+    return arrow::flight::BasicAuth::Deserialize(buf, basic_auth);
+}
+
+Status SerializeBasicAuth(const arrow::flight::BasicAuth &basic_auth, std::unique_ptr<std::string> *out) {
```
Comment (lihalite) [inline cpp/src/arrow/python/flight.cc]: Do we need `unique_ptr<string>`? `string*` should suffice (and that's what the other serialization methods do).

**Review by ghost — COMMENTED** [review]
Looks good to me overall, thanks for doing this! You may also want to run `make format` on the C++ code.

**Comment by rymurr** [comment]
@lihalite how do you get make format to run? I have clang-tidy and clang-format but CMake isn't generating targets for them (apologies for the n00b question!)

**Comment by wesm** [comment]
Do you have the right versions of clang-tidy/clang-format? The targets are only generated if the right versions are found

https://github.com/apache/arrow/blob/master/cpp/CMakeLists.txt#L186

https://github.com/apache/arrow/blob/master/cpp/cmake_modules/FindClangTools.cmake

**Comment by rymurr** [comment]
i figured it out. I was using clang from conda and clang-tidy etc aren't in conda-forge (at least for clang-7). Forcing cmake to use my OS clang-7 worked. Was able to run `make format` now

File: cpp/src/arrow/python/flight.cc (hunk 2 above)
Comment (rymurr) [reply cpp/src/arrow/python/flight.cc]: both have been fixed. Thanks @lihalite

**Comment by pitrou** [comment]
Interesting. I had assumed basic auth used a base64-encode `username:password` string as token.

**Comment by lihalite** [comment]
I think calling it "HTTP basic auth" is misleading, it's really just implementing Flight/Java's "basic" authentication scheme across all languages. (Right now you wouldn't be able to implement HTTP basic auth anyways, as we don't use the Authorization: header)

**Comment by rymurr** [comment]
All builds but appveyor passed and appveyor timed out. Is this something I should re-trigger or is this ok to merge now?

**Comment by pitrou** [comment]
Rebasing from master should hopefully fix the AppVeyor issue.

**Comment by rymurr** [comment]
thanks @pitrou  that did the trick!

**Comment by emkornfield** [comment]
It looks like the build is green, does this need more work or can it be merged?

**Comment by rymurr** [comment]
Hey @emkornfield yes, this is ready to be merged

**Review by wesm — APPROVED** [review]
+1

**Pull request closed without merging**